### PR TITLE
[FAI-286] Implement feature distance scaling

### PR DESCRIPTION
**Pull request by ruivieira**
This PR implements feature distance scaling for counterfactual search when that information is available (specified using `DataDistribution`).
If the feature distribution information is not available, the counterfactual search behaves as previously (using unscaled distances).

### CODE REVIEW

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 1]
```diff
@@ -69,6 +70,24 @@ public CounterfactualExplainer(DataDomain dataDomain, List<Boolean> contraints,
...
+    /**
+     * Create a new {@link CounterfactualExplainer} using OptaPlanner as the underlying engine.
+     *
+     * @param dataDistribution Characteristics of the data distribution as {@link DataDistribution}, if available
+     * @param dataDomain       A {@link DataDomain} which specifies the search space domain
+     * @param contraints       A list specifying by index which features are constrained
+     * @param goal             A collection of {@link Output} representing the desired outcome
+     * @param solverConfig     An OptaPlanner {@link SolverConfig} configuration
+     */
+    public CounterfactualExplainer(DataDistribution dataDistribution, DataDomain dataDomain, List<Boolean> contraints, List<Output> goal, SolverConfig solverConfig, Executor executor) {
```
Comment (danielezonca) [inline explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java:82]: I think we already have something like 6 or more different `CounterfactualExplainer` constructors :)
What about create a nested `Builder` class to collect all params and then in the `build()` method invoke always the same constructor?
Something like [this](https://github.com/kiegroup/drools/blob/master/kie-pmml-trusty/kie-pmml-models/kie-pmml-models-drools/kie-pmml-models-drools-scorecard/kie-pmml-models-drools-scorecard-model/src/test/java/org/kie/pmml/models/drools/scorecard/model/KiePMMLScorecardModelTest.java#L92-L95)

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 2]
```diff
@@ -110,8 +150,13 @@ private Executor getExecutor() {
...
             final Feature feature = predictionInput.getFeatures().get(i);
             final Boolean isConstrained = constraints.get(i);
-            final FeatureDomain featureDistribution = dataDomain.getFeatureDomains().get(i);
-            final CounterfactualEntity counterfactualEntity = CounterfactualEntityFactory.from(feature, isConstrained, featureDistribution);
+            final FeatureDomain featureDomain = dataDomain.getFeatureDomains().get(i);
+            FeatureDistribution featureDistribution = null;
+            // If a data distribution was specified, use it to instantiate the counterfactual entity
+            if (dataDistribution != null) {
+                featureDistribution = dataDistribution.getFeatureDistributions().get(i);
+            }
```
Comment (danielezonca) [inline explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java:158]: Minor style comment
```suggestion
            // If a data distribution was specified, use it to instantiate the counterfactual entity
            final FeatureDistribution featureDistribution = Optional
                .ofNullable(dataDistribution)
                .map(dd -> dd.getFeatureDistributions().get(i))
                .orElse(null);
```

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 1 above)
Comment (ruivieira) [reply explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: very good point :)
implemented.

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 2 above)
Comment (ruivieira) [reply explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: changed.

**Review by danielezonca — APPROVED** [review]
Thanks for your PR, approving with minor comments :+1:

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 3]
```diff
@@ -41,63 +42,40 @@
...
-     * @param goal         A collection of {@link Output} representing the desired outcome
-     * @param solverConfig An OptaPlanner {@link SolverConfig} configuration
+     * @param dataDistribution Characteristics of the data distribution as {@link DataDistribution}, if available
+     * @param dataDomain       A {@link DataDomain} which specifies the search space domain
+     * @param contraints       A list specifying by index which features are constrained
+     * @param goal             A collection of {@link Output} representing the desired outcome
+     * @param solverConfig     An OptaPlanner {@link SolverConfig} configuration
      */
-    public CounterfactualExplainer(DataDomain dataDomain, List<Boolean> contraints, List<Output> goal, SolverConfig solverConfig, Executor executor) {
+    public CounterfactualExplainer(DataDistribution dataDistribution,
```
[the remark this answers is not in the document]
Comment (ruivieira) [reply explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java:64]: done

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java [hunk 4]
```diff
@@ -108,10 +86,18 @@ private Executor getExecutor() {
...
 
         for (int i = 0; i < predictionInput.getFeatures().size(); i++) {
+
             final Feature feature = predictionInput.getFeatures().get(i);
             final Boolean isConstrained = constraints.get(i);
-            final FeatureDomain featureDistribution = dataDomain.getFeatureDomains().get(i);
-            final CounterfactualEntity counterfactualEntity = CounterfactualEntityFactory.from(feature, isConstrained, featureDistribution);
+            final FeatureDomain featureDomain = dataDomain.getFeatureDomains().get(i);
+            // If a data distribution was specified, use it to instantiate the counterfactual entity
+            final int featureIndex = i;
```
[the remark this answers is not in the document]
Comment (ruivieira) [reply explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: Replaced the entire `for loop` with a stream

File: explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ruivieira) [reply explainability/core/counterfactuals/src/main/java/org/kie/kogito/explainability/local/counterfactual/CounterfactualExplainer.java]: constructor now has correct modifier

**Pull request merged**